pkg/apis/policy/v1beta1/conversion.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/kubernetes/pkg/apis/policy"
+)
+
+const (
+	pdbV1beta1Label = "pdb.kubernetes.io/deprecated-v1beta1-empty-selector-match"
+)
+
+func Convert_v1beta1_PodDisruptionBudget_To_policy_PodDisruptionBudget(in *v1beta1.PodDisruptionBudget, out *policy.PodDisruptionBudget, s conversion.Scope) error {
```

Comment:
It's simpler to call the generated auto-conversion method first, which will handle all of the boilerplate, then tweak the selector specifically.

Rather than trying to strip the deprecated v1beta1 label specifically, I _think_ we can do two mechanical transformations in each direction by seeing if the input selector is identical to a specific "all" or "none" selector, and translating that into a new specific selector. Something like this:

```go

var (
	internalMatchAllSelector  = &metav1.LabelSelector{}
	internalMatchNoneSelector = &metav1.LabelSelector{
		MatchExpressions: []metav1.LabelSelectorRequirement{{Key: pdbV1beta1Label, Operator: metav1.LabelSelectorOpExists}},
	}

	v1beta1MatchNoneSelector = &metav1.LabelSelector{}
	v1beta1MatchAllSelector  = &metav1.LabelSelector{
		MatchExpressions: []metav1.LabelSelectorRequirement{{Key: pdbV1beta1Label, Operator: metav1.LabelSelectorOpDoesNotExist}},
	}
)

func Convert_v1beta1_PodDisruptionBudget_To_policy_PodDisruptionBudget(in *v1beta1.PodDisruptionBudget, out *policy.PodDisruptionBudget, s conversion.Scope) error {
	if err := autoConvert_v1beta1_PodDisruptionBudget_To_policy_PodDisruptionBudget(in, out, s); err != nil {
		return err
	}

	switch {
	case apiequality.Semantic.DeepEqual(in.Spec.Selector, v1beta1MatchNoneSelector):
		// If the v1beta1 version has a non-nil but empty selector, it should be
		// selecting no pods, even when used with the internal or v1 api. We
		// add a selector that is non-empty but will never match any pods.
		out.Spec.Selector = internalMatchNoneSelector.DeepCopy()
	case apiequality.Semantic.DeepEqual(in.Spec.Selector, v1beta1MatchAllSelector):
		// If the v1beta1 version has our v1beta1-specific "match-all" selector,
		// swap that out for a simpler empty "match-all" selector for v1
		out.Spec.Selector = internalMatchAllSelector.DeepCopy()
	}
	return nil
}

func Convert_policy_PodDisruptionBudget_To_v1beta1_PodDisruptionBudget(in *policy.PodDisruptionBudget, out *v1beta1.PodDisruptionBudget, s conversion.Scope) error {
	if err := autoConvert_policy_PodDisruptionBudget_To_v1beta1_PodDisruptionBudget(in, out, s); err != nil {
		return err
	}

	switch {
	case apiequality.Semantic.DeepEqual(in.Spec.Selector, internalMatchNoneSelector):
		// If the internal version has our v1beta1-specific "match-none" selector,
		// swap that out for a simpler empty "match-none" selector for v1beta1
		out.Spec.Selector = v1beta1MatchNoneSelector.DeepCopy()
	case apiequality.Semantic.DeepEqual(in.Spec.Selector, internalMatchAllSelector):
		// If the internal version has a non-nil but empty selector, we want it to
		// select all pods. We make sure this happens even with the v1beta1 api by
		// adding a non-empty selector that selects all pods.
		out.Spec.Selector = v1beta1MatchAllSelector.DeepCopy()
	}
	return nil
}
```

Reply:
I have updated the code to first use the auto-conversion and then fix the selector. 

But I think we might need the functionality to strip only the v1beta1 label, since there might be situations we need to convert and the v1beta1 label is not the only element in the selector. Since the selector in PDBs is mutable, I think it would be possible to read a resource as v1beta1, add an additional label to the PDB, and then write it back. If so, we can't rely on the equality checks.

This situation covered by this test: https://github.com/kubernetes/kubernetes/pull/99290/files#diff-ac28b0c739c72c12d59ca4e60b5d43c82c582bc87a005258b2412332d7d72ac3R89.

Reply:
> Since the selector in PDBs is mutable, I think it would be possible to read a resource as v1beta1, add an additional label to the PDB, and then write it back.

I don't think that's an expected use case. Selectors are expected to be completely set by the writer... see https://github.com/kubernetes/kubernetes/issues/92913 and https://github.com/kubernetes/kubernetes/pull/93901

With the all-or-nothing conversion as proposed in #discussion_r589092943, what is the behavior in the following scenarios:

1. empty v1beta1 -> non-empty v1 -> empty v1beta1
    1. `kubectl apply` v1beta1 version of a PDB with an empty selector (`{}`)
    2. `kubectl apply` v1 version of the same PDB with some other selector (`{"matchLabels":{"foo":"bar"}}`) - expect only foo=bar via v1 and v1beta1 on read
    3. `kubectl apply` v1beta1 version of a PDB with an empty selector again (`{}`) - expect no trace of `foo=bar`

2. non-empty v1beta1 -> empty v1 -> non-empty v1beta1
    1. `kubectl apply` v1beta1 version of a PDB with a non-empty selector (`{"matchLabels":{"foo":"bar"}}`) - expect only foo=bar via v1 and v1beta1 on read
    2. `kubectl apply` v1 version of the same PDB with an empty selector (`{}`) - expect no trace of foo=bar
    3. `kubectl apply` v1beta1 version of a PDB with a non-empty selector again (`{"matchLabels":{"foo":"bar"}}`) - expect only foo=bar via v1 and v1beta1 on read

Reply:
for this to behave as expected, I think the selector field has to have patchStrategy=replace

```
diff --git a/staging/src/k8s.io/api/policy/v1/types.go b/staging/src/k8s.io/api/policy/v1/types.go
index 42eab54b3c8..904413203ad 100644
--- a/staging/src/k8s.io/api/policy/v1/types.go
+++ b/staging/src/k8s.io/api/policy/v1/types.go
@@ -34,8 +34,9 @@ type PodDisruptionBudgetSpec struct {
 	// budget.
 	// A nil selector will match no pods, while an empty selector will select
 	// all pods within the namespace.
+	// +patchStrategy=replace
 	// +optional
-	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
+	Selector *metav1.LabelSelector `json:"selector,omitempty" patchStrategy:"replace" protobuf:"bytes,2,opt,name=selector"`
 
 	// An eviction is allowed if at most "maxUnavailable" pods selected by
 	// "selector" are unavailable after the eviction, i.e. even in absence of
diff --git a/staging/src/k8s.io/api/policy/v1beta1/types.go b/staging/src/k8s.io/api/policy/v1beta1/types.go
index 87c827d342b..4a1d2fee596 100644
--- a/staging/src/k8s.io/api/policy/v1beta1/types.go
+++ b/staging/src/k8s.io/api/policy/v1beta1/types.go
@@ -36,8 +36,9 @@ type PodDisruptionBudgetSpec struct {
 	// A nil selector selects no pods. An empty selector also selects no pods, which
 	// is different than the normal behavior for selectors where an empty selector
 	// does not restrict the selection space.
+	// +patchStrategy=replace
 	// +optional
-	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,2,opt,name=selector"`
+	Selector *metav1.LabelSelector `json:"selector,omitempty" patchStrategy:"replace" protobuf:"bytes,2,opt,name=selector"`
 
 	// An eviction is allowed if at most "maxUnavailable" pods selected by
 	// "selector" are unavailable after the eviction, i.e. even in absence of
```

@apelisse @jpbetz - does that match your expectations? is there something similar that needs to be specified for server-side apply to treat this field as atomic, or is that the default?

Reply:
empty_v1beta1.yaml:
```yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
metadata:
  name: empty-v1beta1
spec:
  minAvailable: 2
  selector: {}
```

nonempty_v1.yaml:
```
apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
  name: empty-v1beta1
spec:
  minAvailable: 2
  selector:
    matchLabels:
      foo: bar
```


setting patchStrategy=replace and using a new kubectl aware of the go type change works as expected:

```
$ kubectl apply -f empty_v1beta1.yaml 
$ kubectl get -f empty_v1beta1.yaml  -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector: {}

$ kubectl apply -f nonempty_v1.yaml 
$ kubectl get -f nonempty_v1.yaml -o yaml
apiVersion: policy/v1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector:
    matchLabels:
      foo: bar

$ kubectl apply -f empty_v1beta1.yaml 
$ kubectl get -f empty_v1beta1.yaml  -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector: {}
```

using an old kubectl that doesn't have the go type change compiled in does not work:

```
$ kubectl-1.19 apply -f empty_v1beta1.yaml 
$ kubectl-1.19 get -f empty_v1beta1.yaml -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector: {}

$ kubectl-1.19 apply -f nonempty_v1.yaml 
$ kubectl-1.19 get -f nonempty_v1.yaml -o yaml
apiVersion: policy/v1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector:
    matchExpressions:
    - key: pdb.kubernetes.io/deprecated-v1beta1-empty-selector-match
      operator: Exists
    matchLabels:
      foo: bar

$ kubectl-1.19 apply -f empty_v1beta1.yaml 
$ kubectl-1.19 get -f empty_v1beta1.yaml -o yaml
apiVersion: policy/v1beta1
kind: PodDisruptionBudget
...
spec:
  minAvailable: 2
  selector: {}
```

@apelisse @jpbetz - it looks like we can't change patchStrategy or set patchStrategy=replace on new go types and have old kubectl clients client-side apply correctly?

Reply:
kubectl 1.19 sends `"Content-Type: application/merge-patch+json"` when applying to `policy/v1`:
```json
{
    "metadata": {
        "annotations":{
            "kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"policy/v1\",\"kind\":\"PodDisruptionBudget\",\"metadata\":{\"annotations\":{},\"name\":\"empty-v1beta1\",\"namespace\":\"default\"},\"spec\":{\"minAvailable\":2,\"selector\":{\"matchLabels\":{\"foo\":\"bar\"}}}}\n"
        }
    },
    "spec":{
        "selector":{
            "matchLabels":{"foo":"bar"}
        }
    }
}
```

kubectl build from this branch sends `"Content-Type: application/strategic-merge-patch+json"` when applying to `policy/v1`, and works properly:
```json
{
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"policy/v1\",\"kind\":\"PodDisruptionBudget\",\"metadata\":{\"annotations\":{},\"name\":\"empty-v1beta1\",\"namespace\":\"default\"},\"spec\":{\"minAvailable\":2,\"selector\":{\"matchLabels\":{\"foo\":\"bar\"}}}}\n"
        }
    },
    "spec":{
        "selector":{
            "matchLabels":{"foo":"bar"}
        }
    }
}
```

Reply:
I thought kubectl would read patch info from openapi and send strategic merge patches even on types it did not have compiled in

Reply:
fortunately, both old and new kubectl's behave like a champ with `apply --server-side`

Reply:
looks like kubectl doesn't send strategic merge patches for unknown types, even if the openapi schema indicates they are supported:

https://github.com/kubernetes/kubernetes/blob/4fccba9e068f31a562f753caf4b37959b4ef829a/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L120-L166

I'll open a separate issue to track that, but a fix there will be too late for this to take advantage of. Given that, I think you approach of stripping the deprecated beta label would work well as a default case for the v1beta1->internal conversion (if we don't exactly match one of the well-known selectors). I think we also need to do that check in v1->internal conversion so that when we switch etcd storage to v1 (and never hit this v1beta1 conversion code), we'll catch selectors like that.

Messing around with that in the last commit of https://github.com/liggitt/kubernetes/commits/PromotePDBsToGA seemed to work well.

Reply:
I thought this rang a bell with a problem in _kubecfg_ from two years ago. See https://github.com/bitnami/kubecfg/issues/259#issuecomment-522054301. The related discussion in the "sig-api-machinery" Slack channel starts [here](https://kubernetes.slack.com/archives/C0EG7JC6T/p1565969006291300).

Reply:
Updated based on the suggested solution at https://github.com/liggitt/kubernetes/commits/PromotePDBsToGA